listenbrainz/webserver/static/js/src/RecentListens.test.tsx

```diff
@@ -43,7 +50,353 @@ const props = {
+    const wrapper = shallow<RecentListens>(<RecentListens {...props} />);
+    const instance = wrapper.instance();
+
+    const spy = jest.spyOn(instance["socket"], "emit");
+    instance.handleFollowUserListChange(["foo"]);
+
+    expect(spy).toHaveBeenCalledTimes(1);
+    expect(spy).toHaveBeenCalledWith("json", {
+      user: "iliekcomputers",
+      follow: ["foo"],
+    });
+  });
+
+  it('calls getRecentListensForFollowList if mode is "follow"', () => {
+    const wrapper = shallow<RecentListens>(<RecentListens {...props} />);
+    const instance = wrapper.instance();
+
+    wrapper.setState({ mode: "follow", followList: ["bar"] });
+    instance.getRecentListensForFollowList = jest.fn();
+    instance.handleFollowUserListChange(["foo"]);
+
+    expect(instance.getRecentListensForFollowList).toHaveBeenCalledTimes(1);
+  });
+});
+
+describe("handleSpotifyAccountError", () => {
+  it("calls newAlert", () => {
+    const wrapper = shallow<RecentListens>(<RecentListens {...props} />);
+    const instance = wrapper.instance();
+    instance.newAlert = jest.fn();
+
+    instance.handleSpotifyAccountError(<p>Test</p>);
+    expect(instance.newAlert).toHaveBeenCalledTimes(1);
+    expect(instance.newAlert).toHaveBeenCalledWith(
+      "danger",
+      "Spotify account error",
+      <p>Test</p>
+    );
+  });
+
+  it('sets "canPlayMusic" to false', () => {
+    const wrapper = shallow<RecentListens>(<RecentListens {...props} />);
+    const instance = wrapper.instance();
+
+    wrapper.setState({ canPlayMusic: true });
+
+    instance.handleSpotifyAccountError(<p>Test</p>);
+    expect(wrapper.state().canPlayMusic).toBe(false);
+  });
+});
+
+describe("handleSpotifyPermissionError", () => {
+  it("calls newAlert", () => {
+    const wrapper = shallow<RecentListens>(<RecentListens {...props} />);
+    const instance = wrapper.instance();
+    instance.newAlert = jest.fn();
+
+    instance.handleSpotifyPermissionError(<p>Test</p>);
+    expect(instance.newAlert).toHaveBeenCalledTimes(1);
+    expect(instance.newAlert).toHaveBeenCalledWith(
+      "danger",
+      "Spotify permission error",
+      <p>Test</p>
+    );
+  });
+
+  it('sets "canPlayMusic" to false', () => {
+    const wrapper = shallow<RecentListens>(<RecentListens {...props} />);
+    const instance = wrapper.instance();
+
+    wrapper.setState({ canPlayMusic: true });
+
+    instance.handleSpotifyPermissionError(<p>Test</p>);
+    expect(wrapper.state().canPlayMusic).toBe(false);
+  });
+});
+
+describe("receiveNewListen", () => {
+  const mockListen: Listen = {
+    track_metadata: {
+      artist_name: "Coldplay",
+      track_name: "Viva La Vida",
+    },
+    listened_at: 1586580524,
+    listened_at_iso: "2020-04-10T10:12:04Z",
+  };
+
+  it("crops the listens array if length is more than or equal to 100", () => {
+    const wrapper = shallow<RecentListens>(
+      <RecentListens
+        {...(JSON.parse(JSON.stringify(tooManyListens)) as RecentListensProps)}
```

Comment:
do we need the json.parse(json.dumps(object)) here?

Reply:
Yes, we do, otherwise  `tooManyListens` gets passed by reference. Which means `result` and and the `state` will always be the same. `JSON.parse(HSON.stringify(object))` is a fast way to deep copy so that this doesn't occur.

Reply:
Cool, could you add a comment explaining that? So that the next person who sees the code understands it.